### DISCUSSION TITLE
How to Wait on Flux Reconcilliation

### DISCUSSION THREAD

**Discussion by jcristol** [fluxcd/flux2#2320] [Q&A] [answered]
Hi!

I've been using flux at my org for a while and have built a workflow in fluxv1 that goes as follows.

All of these are happening inside a ci/cd system as subsequent steps.
1. make a change to our flux repo for a rolling out a new version of a deployment. i.e. bumping the version on a docker image
2. explicitly wait for flux to sync the image update to the cluster
  - this involves running fluxctl sync etc.
  - and run kubectl get deployment etc. to ensure that the deployment exists in the cluster, the deployment is in the ready state, and is at the latest image version 
3. run some automated smoke tests against the new deployment version

Step 2 is something that we don't have out of the box in fluxv1 from what I can tell. Do we have a new flux api in v2 that would allow us to confirm as part of a ci pipeline whether the git changes were applied successfully to the cluster and the cluster reflects the new state in git?

Hope this makes sense I can elaborate further if needed.

**Comment by kingdonb** [comment, marked as answer]
> 1. make a change to our flux repo for a rolling out a new version of a deployment
> 2. explicitly wait for flux to sync the image update to the cluster
> 3. run some automated smoke tests against the new deployment version

What you're describing is the primary use of Flagger, another project in the Flux family.

https://docs.flagger.app/usage/how-it-works

Flagger tracks the status of an original deployment target (the "canary deployment" now) which receives some subset of traffic during the rollout testing, and is promoted in a controlled way, that you define.

You can run Helm Test or other smoke tests as a pre-rollout hook:

https://docs.flagger.app/usage/webhooks

As for doing (2) with Flux alone, we don't have an API for "waiting until X commit hash is deployed" that I am aware of, but it would be easy enough to call `flux get kustomization <ks-to-watch>` in a loop until the status reflects that everything is OK and at the expected commit hash, or waiting until a time out or error condition is reached. The `flux reconcile kustomization <ks-to-sync>` command will do 80% of this for you, you can even adjust `--timeout` in the CLI to compensate for longer wait times, it will return success or failure after that time regardless of what state the Kustomization is in server-side, (even if it's still working – the timeouts are independent.)

My suggestion - use `spec.wait` + notifications if you want to avoid writing a retry-timeout loop and controlling the assessment logic on your own, since Canary does that, you can control the process explicitly and in detail through Flagger directives.

If you don't want to use Flagger, then `flux reconcile` from the command line + setting `spec.wait` in the Kustomization, or explicit health assessment checks of the particular deployments of interest are the way to go. Hope this helps.
